enhancements/network/persistent-ips-sdn-secondary-networks.md

```diff
@@ -0,0 +1,1016 @@
+
+  CNI->>CNI: ipamClaim := readIPAMClaim(claimName)
+  CNI->>CNI: configIface(ipamClaim.Status.IPs)
+
+  CNI-->>Kubernetes: OK
+
+  Kubernetes-->>KubeVirt: OK
+
+
+  KubeVirt-->>user: OK
+```
+
+When creating the pod, the OVN-Kubernetes IPAM module finds existing
+`IPAMClaim`s for the workload. It will thus use those already reserved
+allocations, instead of generating brand new allocations for the pod where the
+encapsulating object will run. The migration scenario is similar.
+
+#### Removing a Virtual Machine
+This flow is - from a CNI perspective - quite similar to the
+[Stopping a VM flow](#stopping-a-virtual-machine). The main difference is after
+the VM is deleted, Kubernetes Garbage Collection will kick in, and remove the
+orphaned `IPAMClaim`s. OVN-Kubernetes will need to react to the `IPAMClaim`
+delete event, to return those IP addresses to the respective IP pools.
+
+```mermaid
+sequenceDiagram
+
+  actor user
+  participant KubeVirt
+  participant Kubernetes
+  participant CNI
+  participant IPAM CNI
+
+  user->>KubeVirt: deleteVM(vmName)
+
+  KubeVirt->>Kubernetes: deletePod(launcher-vmName)
+
+  Kubernetes->>CNI: CNI DEL
+  CNI-->>Kubernetes: OK
+
+  Kubernetes-->>KubeVirt: OK
+
+  KubeVirt-->>user: OK
+
+  loop for every orphaned IPAMClaim
+    Kubernetes->>Kubernetes: deleteIPAMClaim()
+  end
+
+  note over IPAM CNI: this CNI has an IPAMClaim informer
+
+Kubernetes->>IPAM CNI: removeEvent(ipamClaim)
+IPAM CNI->>IPAM CNI: returnToPool(ipamClaim.Status.IPs)
+```
+
+#### Hot-plug a VM interface
+This flow is exactly the same as
+[starting a VM](#starting-a-previously-stopped-virtual-machine).
+
+#### Hot-unplug a VM interface
+Hot-unplugging an interface from a KubeVirt VM is a two step process, requiring
+a migration:
+1. the user mutates the VM.Spec.Template of the running VMI setting the
+   interface(s) as absent
+2. the user migrates the VM
+3. the interfaces marked as absent will not be templated on the destination pod
+   (i.e. the migration destination pod will not have those interfaces)
+4. KubeVirt dettaches the interface from the live VM
+
+To ensure the IPs allocated to hot-unplugged interfaces are returned to the
+pool, KubeVirt needs to delete the `IPAMClaim`, otherwise, they will only be
+garbage collected when the VM is deleted.
+
+Finally, OVN-Kubernetes must react to the deletion of the IPAMClaim, and return
+this IP address to its in-memory pool.
+
+### API Extensions
+
+We plan on adding a new CRD that would be hosted in the k8snetworkplumbingwg,
+and hopefully accepted as part of the
+[Kubernetes multi-networking de-facto standard](https://github.com/k8snetworkplumbingwg/multi-net-spec).
+Meaning we envision other plugins with IPAM functionality to use this CRD to
+provide persistent IPs for their workloads. This proposal is captured in the
+following
+[de-facto standard update proposal](https://docs.google.com/document/d/1HjPZCfl_3zsXr905T620mF7OFt2QnwxLZomPbeDj0vc).
```

Comment:
Is this proposal tied to a review/approval process? Is that process completion a requirement for this enhancement?